### PR TITLE
feat: Add Inbound and Outbound accounts endpoints

### PR DESCRIPTION
**Pull request by sridharvoruganti**
partly closes #[1999](https://github.com/mojaloop/project/issues/1999)

- implement state machine to handle inbound and outbound calls
- GET `/accounts/{id}`
- PUT `/accounts/{id}`
- PUT `/accounts/{id}/error`

### CODE REVIEW

File: src/handlers/inbound/accounts/{ID}.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  InboundAccountsModelConfig
+} from '~/models/inbound/accounts.model'
+import {
+  OutboundAccountsModel
+} from '~/models/outbound/accounts.model'
+import { Request, ResponseObject } from '@hapi/hapi'
+import { StateResponseToolkit } from '~/server/plugins/state'
+
+/**
+ * Handles a inbound PUT /accounts/{ID} request
```
Comment (eoln) [inline src/handlers/inbound/accounts/{ID}.ts:44]:
```suggestion
 * Handles an inbound PUT /accounts/{ID} request
```

File: src/handlers/inbound/accounts/{ID}.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+async function put (_context: any, request: Request, h: StateResponseToolkit): Promise<ResponseObject> {
```
Comment (eoln) [inline src/handlers/inbound/accounts/{ID}.ts:47]: can't we use `unknown` instead of `any`

File: src/handlers/inbound/accounts/{ID}.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  const userId: string = request.params.ID
+  const channel = OutboundAccountsModel.notificationChannel(userId)
+  const pubSub = h.getPubSub()
+  // don't await on promise to resolve, let finish publish in background
+  pubSub.publish(channel, payload as unknown as Message)
+  return h.response().code(200)
+}
+
+/**
+ * Handles a inbound GET /accounts/{ID} request
```
Comment (eoln) [inline src/handlers/inbound/accounts/{ID}.ts:58]:
```suggestion
 * Handles an inbound GET /accounts/{ID} request
```

File: src/handlers/inbound/accounts/{ID}.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  const userId: string = request.params.ID
+  const logger = h.getLogger()
+  const sourceFspId = request.headers['fspiop-source']
+  const config: InboundAccountsModelConfig = {
+    logger,
+    backendRequests: h.getBackendRequests(),
+    thirdpartyRequests: h.getThirdpartyRequests()
+  }
+  const model = new InboundAccountsModel(config)
+  const response = await model.getUserAccounts(userId, sourceFspId)
```
Comment (eoln) [inline src/handlers/inbound/accounts/{ID}.ts:72]: this is blocking Switch until model.getUserAccounts promise is resolved
setImmediate should be used to not block `h.response().code(202)`

```suggestion
setImmediate(async () => {
  const response = await model.getUserAccounts(userId, sourceFspId)
  logger.push({ response }).info('InboundAccountsModel handled GET /accounts/{ID} request')
})
```

File: src/interface/inbound/api_template/paths/accounts_ID.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
+parameters:
```
Comment (eoln) [inline src/interface/inbound/api_template/paths/accounts_ID.yaml:1]: maybe this inbound path and corresponding error are good candidates to be moved to https://github.com/mojaloop/api-snippets/tree/master/thirdparty/openapi3/paths ?

File: src/models/outbound/accounts.model.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    const pubSub: PubSub = this.pubSub
+
+    // eslint-disable-next-line no-async-promise-executor
+    return new Promise(async (resolve, reject) => {
+      let subId = 0
+      try {
+        // in handlers/inbound is implemented UpdateAccountsByUserId handler
+        // which publish getAccounts response to channel
+        subId = this.pubSub.subscribe(channel, async (channel: string, message: Message, sid: number) => {
+          // if (!message) {
```
Comment (eoln) [inline src/models/outbound/accounts.model.ts:102]: drop this commented code

File: test/unit/models/inbound/accounts.model.test.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    test('reformating of thrown exception when no user accounts returned', async () => {
+      mocked(config.backendRequests.getUserAccounts).mockImplementationOnce(() => Promise.resolve())
+
+      await model.getUserAccounts(userId, dfspId)
+
+      expect(config.backendRequests.getUserAccounts).toHaveBeenCalledWith(userId)
+      expect(config.thirdpartyRequests.putAccountsError).toHaveBeenCalledWith(
+        userId,
+        {
+          errorInformation: {
```
Comment (eoln) [inline test/unit/models/inbound/accounts.model.test.ts:190]: shouldn't it be a generic ID not-found error? 3200

File: test/unit/models/outbound/accounts.model.test.ts [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+
+describe('OutboundAccountsModel', () => {
+  const connectionConfig: RedisConnectionConfig = {
+    port: 6789,
+    host: 'localhost',
+    logger: mockLogger()
+  }
+  let modelConfig: OutboundAccountsModelConfig
+  const expectedResp = {
+    "accounts": {
```
Comment (eoln) [inline test/unit/models/outbound/accounts.model.test.ts:67]: shouldn' the `accounts` be an array ?

**Review by eoln — CHANGES_REQUESTED** [review]
- some typos, 
- there is a strange `accounts` expected value in uni tests where `accounts` is the number indexed object instead of an array,
- maybe is worth waiting on @kleyow  until he finishes the migration of the `Async2Sync` model generator to implement Outbond model ?

File: src/handlers/inbound/accounts/{ID}.ts (hunk 2 above)
Comment (sridharvoruganti) [reply src/handlers/inbound/accounts/{ID}.ts]: ok

File: src/handlers/inbound/accounts/{ID}.ts (hunk 4 above)
Comment (sridharvoruganti) [reply src/handlers/inbound/accounts/{ID}.ts]: Fixed

File: src/interface/inbound/api_template/paths/accounts_ID.yaml (hunk 5 above)
Comment (sridharvoruganti) [reply src/interface/inbound/api_template/paths/accounts_ID.yaml:1]: I have defined all in local paths, may be we can move all those to `api-snippets` like `sdk-scheme-adapter` in a separate ticket.

Comment (kleyow) [reply src/interface/inbound/api_template/paths/accounts_ID.yaml:1]: Indeed. Lets leave it for now. We can refactor paths once complete thirdparty API definitions#2089
is done.

File: test/unit/models/outbound/accounts.model.test.ts (hunk 8 above)
Comment (sridharvoruganti) [reply test/unit/models/outbound/accounts.model.test.ts]: fixed

**Comment by kleyow** [comment]
> maybe is worth waiting on @kleyow until he finishes the migration of the Async2Sync model generator to implement Outbond model ?

I should have this ready when you start working on Friday so you can hold off on the outbound.

**Comment by sridharvoruganti** [comment]
> > maybe is worth waiting on @kleyow until he finishes the migration of the Async2Sync model generator to implement Outbond model ?
> 
> I should have this ready when you start working on Friday so you can hold off on the outbound.

All other outbound API's using existing Model right, so if require I will incorporate Async2Sync refactor changes to Accounts model in followup PR, it's blocking e2e tests.

**Comment by kleyow** [comment]
No problem. Either is fine.

> > > maybe is worth waiting on @kleyow until he finishes the migration of the Async2Sync model generator to implement Outbond model ?
> > 
> > 
> > I should have this ready when you start working on Friday so you can hold off on the outbound.
> 
> All other outbound API's using existing Model right, so if require I will incorporate Async2Sync refactor changes to Accounts model in followup PR, it's blocking e2e tests.

**Review by eoln — APPROVED** [review]

File: test/unit/models/inbound/accounts.model.test.ts (hunk 7 above)
Comment (sridharvoruganti) [reply test/unit/models/inbound/accounts.model.test.ts:198]: Fixed,  previously checking empty response, but ideally simulator returns 3200 error if user id not found

**Pull request merged**